Sources/Tools/plFilePatcher/main.cpp

```diff
@@ -0,0 +1,186 @@
+}
+
+static size_t consoleWidth = 0;
+static ST::string lastProgress;
+
+static void FileDownloadBegin(const plFileName& file)
+{
+    // To account for terminal resizing, we want to refresh this value
+    // periodically, but not every time we draw to the screen, so let's grab it
+    // whenever we start downloading a new file (which is probably still
+    // excessive tbh)
+    GetConsoleWidth(consoleWidth);
+
+    size_t fill = consoleWidth - file.GetSize();
+    ST::printf("\r{}{}\n{}", file, ST::string::fill(fill, ' '), lastProgress);
+}
+
+static void FileDownloadProgress(uint64_t bytesDown, uint64_t bytesTotal, const ST::string& stats)
+{
+    size_t availWidth = consoleWidth - stats.size() - 3;
+    size_t progWidth = size_t(double(bytesDown) / double(bytesTotal) * availWidth);
+
+    lastProgress = ST::format("|{}{}| {}", ST::string::fill(progWidth, '='), ST::string::fill(availWidth - progWidth, ' '), stats);
+    ST::printf("\r{}", lastProgress);
+}
+
+int main(int argc, char* argv[])
+{
+    enum { kArgServerIni, kArgDataOnly, kArgClientOnly, kArgQuiet, kArgHelp1, kArgHelp2 };
+    const plCmdArgDef cmdLineArgs[] = {
+        { kCmdArgFlagged | kCmdTypeString,  "ServerIni",    kArgServerIni },
+        { kCmdArgFlagged | kCmdTypeBool,    "DataOnly",     kArgDataOnly },
+        { kCmdArgFlagged | kCmdTypeBool,    "ClientOnly",   kArgClientOnly },
+        { kCmdArgFlagged | kCmdTypeBool,    "quiet",        kArgQuiet },
+        { kCmdArgFlagged | kCmdTypeBool,    "help",         kArgHelp1 },
+        { kCmdArgFlagged | kCmdTypeBool,    "?",            kArgHelp2 },
+    };
+
+    std::vector<ST::string> args;
+    args.reserve(argc);
+    for (size_t i = 0; i < argc; i++) {
+        args.emplace_back(ST::string::from_utf8(argv[i]));
+    }
+
+    plCmdParser cmdParser(cmdLineArgs, std::size(cmdLineArgs));
+    if (cmdParser.Parse(args)) {
+        if (cmdParser.GetBool(kArgHelp1) || cmdParser.GetBool(kArgHelp2)) {
+            plFileName exeName = plFileName(cmdParser.GetProgramName());
+            PrintUsage(exeName.GetFileName());
+            return 0;
+        }
+    } else {
+        ST::printf(stderr, "An error occurred while parsing the provided arguments.\n");
+        ST::printf(stderr, "Use the --help option to display usage information.\n");
+        return 1;
+    }
+
+    plFileName serverIni = ST_LITERAL("server.ini");
+    if (cmdParser.IsSpecified(kArgServerIni))
+        serverIni = cmdParser.GetString(kArgServerIni);
+
+    plFilePatcher patcher(serverIni);
+    uint32_t flags = plFilePatcher::kPatchEverything;
+
+    if (cmdParser.GetBool(kArgDataOnly))
+        flags &= plFilePatcher::kPatchData;
+    if (cmdParser.GetBool(kArgClientOnly))
+        flags &= plFilePatcher::kPatchClient;
+
+    // The options aren't mutually exclusive, so if we've ended up telling it
+    // not to patch anything, we'll set it back to patching everything.
+    if (!flags)
+        flags = plFilePatcher::kPatchEverything;
```

Comment:
Considering that the options aren't mutually exclusive, I think I would prefer to see the flags start at zero and the command line flags perform a bitwise OR. The logic is more clear to me in that way.